docs: Add Dagger Shell to features list (#9893)
dagger/dagger: docs/current_docs/features/shell.mdx

Dagger Shell enables you to compose pipelines on the fly, by chaining together core Dagger primitives like `Container` and `Directory` with module-defined types and functions. It can be used inline, with scripts, or interactively. It is secure by design, as all commands execute in containers and you have complete control over what goes in or out from the host system.

Here's an example of Dagger Shell in action:

<Tabs groupId="shell">
<TabItem value="System shell">
```shell
dagger <<EOF
container \
  | from alpine \
  | with-exec apk add curl \
  | with-exec -- curl -L https://dagger.io \
  | stdout
EOF
```
</TabItem>
<TabItem value="Dagger Shell">
```shell title="Type 'dagger' to start Dagger Shell in interactive mode and enter the command below."
container | from alpine | with-exec apk add curl | with-exec -- curl -L https://dagger.io | stdout
```
</TabItem>
</Tabs>

![Dagger Shell](/img/current_docs/features/shell-curl.gif)

Here's another, more complex example:

<Tabs groupId="shell">
<TabItem value="System shell">
```shell
dagger <<EOF

> container \
>   | from cgr.dev/chainguard/wolfi-base \
>   | with-exec apk add go \
>   | with-directory /src https://github.com/golang/example#master \
>   | with-workdir /src/hello \
>   | with-exec -- go build -o hello . \
>   | file ./hello \
>   | export ./hello-from-dagger

Reviewer: ```suggestion
container |
  from cgr.dev/chainguard/wolfi-base |
  with-exec apk add go |
  with-directory /src https://github.com/golang/example#master |
  with-workdir /src/hello |
  with-exec -- go build -o hello . |
  file ./hello |
  export ./hello-from-dagger
```

EOF
```
</TabItem>
<TabItem value="Dagger Shell">
```shell title="Type 'dagger' to start Dagger Shell in interactive mode and enter the command below."

> container \
>   | from cgr.dev/chainguard/wolfi-base \
>   | with-exec apk add go \
>   | with-directory /src https://github.com/golang/example#master \
>   | with-workdir /src/hello \
>   | with-exec -- go build -o hello . \
>   | file ./hello \
>   | export ./hello-from-dagger

Reviewer: ```suggestion
container |
  from cgr.dev/chainguard/wolfi-base |
  with-exec apk add go |
  with-directory /src https://github.com/golang/example#master |
  with-workdir /src/hello |
  with-exec -- go build -o hello . |
  file ./hello |
  export ./hello-from-dagger
```

:::

## Modes of execution

Dagger Shell supports multiple ways to run commands:

<Tabs groupId="shell">
<TabItem value="Inline execution">
```shell
dagger -c 'container | from alpine | terminal'
```
</TabItem>
<TabItem value="Standard input">
```shell title="Type 'dagger' to start Dagger Shell in interactive mode and enter the command below."
echo 'container | from alpine | terminal' | dagger
```
</TabItem>
<TabItem value="Script">
```shell
echo 'container | from alpine | with-exec cat /etc/os-release | stdout' > script.sh
dagger script.sh
```
</TabItem>
<TabItem value="Interactive REPL">
```shell title="Type 'dagger' to start Dagger Shell in interactive mode and enter the command below."
container | from alpine | terminal
```
</TabItem>
</Tabs>

## Built-in help

Dagger Shell provides extensive built-in help:

```shell
.help               # Shows available commands
.help <module>      # Shows description, entrypoint arguments, and available functions
.help <function>    # Shows description, arguments, and return type for function

> <function> | .help  # Shows available functions for object returned by <function>

Reviewer: Sorry, I think this is better:
```suggestion
<function> | .help  # Shows details on the type returned by <function>, including available functions if an object
```

:::note
To safeguard access to the host filesystem, the context root cannot be escaped.
:::

## Modules

If the working directory contains a Dagger module, that module is automatically loaded, and its Dagger Functions can be directly inspected (via `.help`) or invoked.

![Dagger Shell module loading](/img/current_docs/features/shell-module.gif)

:::tip
If only the core Dagger API is needed, the `-n` (`--no-mod`) flag can be provided. This reduces the startup time because the Dagger CLI doesn't try to find and load a current module.
:::

## Paths

Filesystem paths are handled differently depending on the initial context of the module loaded in Dagger Shell.

- If the loaded module is local, or if no module is loaded, directories or files will come from the host system.
- If the loaded module is remote, directories or files will come from that remote directory.
  - In this latter case, to use files and directories from the host system, they must be referred to using the `host` function.

Here's an example:

<Tabs groupId="shell">
<TabItem value="System shell">
```shell
dagger <<EOF

> github.com/dagger/dagger/modules/wolfi@v0.16.2 \
>   | container \
>   | with-file /README.md $(host | file ./README.md) \
>   | with-exec cat /README.md \
>   | stdout

Reviewer: ```suggestion
github.com/dagger/dagger/modules/wolfi@v0.16.2 |
  container |
  with-file /README.md $(host | file ./README.md) |
  with-exec cat /README.md |
  stdout
```

EOF
```
</TabItem>
<TabItem value="Dagger Shell">
```shell title="Type 'dagger' to start Dagger Shell in interactive mode and enter the command below."

> github.com/dagger/dagger/modules/wolfi@v0.16.2 \
>   | container \
>   | with-file /README.md $(host | file ./README.md) \
>   | with-exec cat /README.md \
>   | stdout

Reviewer: ```suggestion
github.com/dagger/dagger/modules/wolfi@v0.16.2 |
  container |
  with-file /README.md $(host | file ./README.md) |
  with-exec cat /README.md |
  stdout
```

:::

## Modes of execution

Dagger Shell supports multiple ways to run commands:

<Tabs groupId="shell">
<TabItem value="Inline execution">
```shell
dagger -c 'container | from alpine | terminal'
```
</TabItem>
<TabItem value="Standard input">
```shell title="Type 'dagger' to start Dagger Shell in interactive mode and enter the command below."
echo 'container | from alpine | terminal' | dagger
```
</TabItem>
<TabItem value="Script">
```shell
echo 'container | from alpine | with-exec cat /etc/os-release | stdout' > script.sh
dagger script.sh
```
</TabItem>
<TabItem value="Interactive REPL">
```shell title="Type 'dagger' to start Dagger Shell in interactive mode and enter the command below."
container | from alpine | terminal
```
</TabItem>
</Tabs>

## Built-in help

Dagger Shell provides extensive built-in help:

```shell

> .help             # Shows available commands
> .help MODULE      # Shows description, arguments, and return type for module constructor
> .help FUNCTION    # Shows description, arguments, and return type for function
> MODULE | .help    # Shows available functions for constructor returned by MODULE
> FUNCTION | .help  # Shows available functions for object returned by FUNCTION

Reviewer: Note that our CLI use the `<module>` convention rather than `MODULE` when describing usage syntax.

```suggestion
.help               # Shows available commands
.help <module>      # Shows description, entrypoint arguments, and available functions
.help <function>    # Shows description, arguments, and return type for function
<function> | .help  # Shows available functions for object returned by <function>
```

Here's an example of Dagger Shell in action:

![Dagger Shell](/img/current_docs/features/shell-curl.gif)

Here is another, more complex example:

![Dagger Shell](/img/current_docs/features/shell-build.gif)

:::note
Dagger Shell allows you to execute Dagger Functions as native shell commands, using a Bash-compatible syntax. However, instead of calling a subprocess for execution, Dagger Shell directly interprets commands within the Dagger CLI and executes them in containers. This means that all your commands execute as API calls in a secure, sandboxed environment, rather than as processes with direct access to the host system.
:::

## Modes of execution

Dagger Shell supports multiple ways to run commands:

<Tabs groupId="shell">
<TabItem value="Inline execution">
```shell
dagger -c 'container | from alpine | terminal'
```
</TabItem>
<TabItem value="Standard input">
```shell title="Type 'dagger' to start Dagger Shell in interactive mode and enter the command below."
echo 'container | from alpine | terminal' | dagger
```
</TabItem>
<TabItem value="Script">
```shell

> dagger script.sh

Author: Is there an example of this script syntax?

Reviewer: What do you mean? It's the same as what you pass in `-c` or stdin, just provided as a file.

Reviewer: If you want a self-contained example:
```shell
echo 'container | from alpine | with-exec cat /etc/os-release | stdout' > script.sh
dagger script.sh
```

```
</TabItem>
</Tabs>

## Built-in help

Dagger Shell provides extensive built-in help:

```shell
.help             # Shows available commands
.help MODULE      # Shows description, arguments, and return type for module constructor
.help FUNCTION    # Shows description, arguments, and return type for function
MODULE | .help    # Shows available functions for constructor returned by MODULE
FUNCTION | .help  # Shows available functions for object returned by FUNCTION
```

![Dagger Shell help](/img/current_docs/features/shell-help.gif)

## Navigation

Dagger Shell presents a virtual filesystem to the user. When Dagger Shell starts, it sets the working directory to an initial context, which defaults to the enclosing Git repository or Dagger module. The user can navigate the virtual filesystem by changing the working directory.

The initial context can be a local filesystem path or a remote filesystem. For example, `dagger -m github.com/dagger/dagger/docs@v0.13.0` changes the context to `github.com/dagger/dagger@v0.13.0` and the working directory to `/docs`.

:::note
To safeguard access to the host filesystem, the context root cannot be escaped.
:::

## Modules

> If the working directory contains a Dagger module, that module is automatically loaded, and its Dagger Functions can be directly inspected (via `.help`) or invoked.

Reviewer: It's worth mentioning that if only the core API is needed, the `-n` (`--no-mod`) flag can be provided which makes for a faster load because the CLI doesn't try to find and load the current module. This replaces `dagger core`.

If `-n` nor `-m` are provided and the current workdir is not in a module, it will fallback to loading the core API only. So it's like a `dagger call` that fallsback to `dagger core`.
